fix(flags): Use a custom signal to trigger the activity log

haacked opened the pull request:
This adds a new custom signal: `model_activity_signal` (in `signals.py`) and a new mixin `ModelActivityMixin`. Any model that has this mixin in its inheritance tree will automatically raise the `model_activity_signal` with a `before` and `after` instance.

Then to log the activity, all you have to do is add a handler method to handle the signal by decorating a method with: `@receiver(model_activity_signal, sender={ModelType})` replacing `{ModelType}` with your model type.

For example, here's what it looks like for `FeatureFlag` in `posthog/api/feature_flag.py`:

```python
@receiver(model_activity_signal, sender=FeatureFlag)
def handle_feature_flag_change(sender, scope, before_update, after_update, activity, was_impersonated=False, **kwargs):
    log_activity(
        organization_id=after_update.team.organization_id,
        team_id=after_update.team_id,
        user=after_update.last_modified_by,
        was_impersonated=was_impersonated,
        item_id=after_update.id,
        scope=scope,
        activity=activity,
        detail=Detail(
            changes=changes_between(scope, previous=before_update, current=after_update), name=after_update.key
        ),
    )
```

Right now, only `FeatureFlag` uses this, but I hope others can adapt this to their needs. We could probably do even more in the Mixin so it's even easier for other models to take advantage of this pattern, but I think for now this works well.

/cc @pauldambra curious to hear your thoughts about this approach. I think it might help other models adapt to having an activity log.

Also, if we like this approach, I wouldn't mind refactoring other models to use it as a follow-up PR.

## Problem

We couldn't be sure that all changes to a feature flag were definitively logged in the activity log. This makes debugging customer issues challenging. For example, scheduled changes were not reflected in the activity log.

Fixes #29448
Fixes #29606

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Updated unit tests and manually tested the code:

### Test Plan

Use two different users for each test.

- [ ] Test scheduled change of a feature flag (_found existing issue with dispatch. Will fix as part of this._)
- [x] Test creating a flag
- [x] Test updating a flag
- [x] Test creating a flag via an experiment
- [x] Test updating a flag via an experiment
- [x] Test creating a flag as an impersonated user
- [x] Test updating a flag as an impersonated user
- [x] Test creating a flag via an experiment as an impersonated user
- [x] Test updating a flag via an experiment as an impersonated user

haacked committed:
Don't create a flag in the test setup

That's going to trigger an activity log entry now, whereas before, it didn't.

haacked committed:
Use a custom signal to trigger the activity log

This adds a new custom signal: `model_activity_signal` and a mixin `ModelActivityMixin`. Any model that has this mixin in its inheritance tree will automatically raise the `model_activity_signal` with a `before` and `after` instance.

Then to log the activity, all you have to do is add a handler method to handle the signal by decorating a method with: `@receiver(model_activity_signal, sender=FeatureFlag)` replacing `FeatureFlag` with your model type.

Right now, only FeatureFlag uses this, but I hope others can adapt this to their needs.

Fixes #29448
Fixes #29606

haacked committed:
Fix serialization error

Need to skip these when serializing a flag for the activity log diff.

haacked commented:
Some implementation notes:

The reason I couldn't use the built-in `post_save` or `pre_save` signals is they don't give you access to the before (or after). Only the current state.

The reason I couldn't just override the `save` method on the model is it lead to circular imports. I could have probably spent a lot of time teasing those apart, but creating a custom signal decoupled the code nicely and made it easy to complete the implementation.

haacked commented on posthog/api/feature_flag.py:
```diff
@@ -1083,41 +1096,21 @@ def activity(self, request: request.Request, **kwargs):
-            activity="updated",
-            detail=Detail(changes=changes, name=serializer.instance.key),
-        )
+@receiver(model_activity_signal, sender=FeatureFlag)
+def handle_feature_flag_change(sender, scope, before_update, after_update, activity, was_impersonated=False, **kwargs):
+    log_activity(
+        organization_id=after_update.team.organization_id,
+        team_id=after_update.team_id,
+        user=after_update.last_modified_by,
+        was_impersonated=was_impersonated,
+        item_id=after_update.id,
+        scope=scope,
+        activity=activity,
+        detail=Detail(
+            changes=changes_between(scope, previous=before_update, current=after_update), name=after_update.key
+        ),
+    )
```
??? What code is being duplicated?

haacked replied:
Maybe it's being confused by the diff? Because this actually reduced duplication of the `log_activity` calls.

pauldambra commented on posthog/api/feature_flag.py:
```diff
@@ -1083,41 +1096,21 @@ def activity(self, request: request.Request, **kwargs):
+    log_activity(
+        organization_id=after_update.team.organization_id,
+        team_id=after_update.team_id,
+        user=after_update.last_modified_by,
```
ah... this was going to be my question but you have the acting user captured on your model 💅 

i think this is great... keeps the two concerns separate

the activity log isn't bound directly to django model so it stays as flexible as today
but folk whose django model carries all the information they need can have a simpler implementation

haacked replied:
> the activity log isn't bound directly to django model so it stays as flexible as today but folk whose django model carries all the information they need can have a simpler implementation

Yup! I imagine that the activity log isn't bound to a specific model, but it's probably usually triggered by a model change. So in theory you could have code that listens to multiple models and decides how to write the activity log. This'll make the listening part easier in those scenarios.

pauldambra commented on posthog/models/activity_logging/model_activity.py:
```diff
@@ -0,0 +1,54 @@
+        change_type = "updated" if self.pk else "created"
+
+        # Get request from thread local
+        request = get_current_request()
```
i guess folk could forget to set this (or not know it's required) and we should give some helpful error if they do

haacked replied:
Yeah, good call. This is the most brittle part of this approach. But I noticed that feature flag scheduled changes set the request user correctly, so we must be doing this sort of thing elsewhere too.

haacked replied:
Actually, it's only required if you were impersonating. I think I should make it more explicit by making you set `was_impersonated` in tls rather than passing the `request`.

pauldambra reviewed (commented):
didn't run it/test it

so just approving the general approach and will let you all decide (pun intended) on the PR itself 😊

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

haacked committed:
Add an ignore for mypy's sake

havenbarnes commented on .github/workflows/ci-backend.yml:
```diff
@@ -142,6 +142,7 @@ jobs:
 
             - name: Check static typing
               run: |
+                  mypy --version  
```
guessing this was for testing mypy discrepanies, just a reminder to remove unless you think it's helpful to leave!

havenbarnes commented on posthog/models/activity_logging/model_activity.py:
```diff
@@ -0,0 +1,63 @@
+    return getattr(_thread_local, "was_impersonated", False)
+
+
+class ModelActivityMixin(models.Model):
```
This is really cool, imo worth sharing at all hands next week or in #dev at least!

havenbarnes commented on posthog/api/feature_flag.py:
```diff
@@ -1130,41 +1134,21 @@ def activity(self, request: request.Request, **kwargs):
-            detail=Detail(changes=changes, name=serializer.instance.key),
-        )
+@receiver(model_activity_signal, sender=FeatureFlag)
+def handle_feature_flag_change(sender, scope, before_update, after_update, activity, was_impersonated=False, **kwargs):
```
nit: Does `handle_feature_flag_change` belong in this file? Since it would capture changes made directly to the model still without going through a FF API, right?

haacked replied:
Correct, however this can't be in `poshog/models/feature_flag.py` because of circular imports, so I thought this is the other place where we make changes to `FeatureFlag`. It "conceptually" felt good to put this here.

Another option is for me to add a new file, perhaps next to `posthog/models/feature_flag.py` that I might call `feature_flag_extensions.py` or `feature_flag_change_handler` and move this method there. Might make sense to also move `scheduled_changes_dispatcher` into that file too. That way we group the ways that `FeatureFlag` can change into a single place.

Let me know what you think!

havenbarnes reviewed (approved):
Looks great! Just left the one nit

haacked committed:
Use the correct user for the change

When scheduling a change, we want the user who created the scheduled change to be the one who modifies the feature flag.

haacked committed:
Add type hint

haacked committed:
Update mypy-baseline.txt
